File: funky_fractions_multi.c
#include <stdio.h>
#include <stdlib.h>
#include <time.h>
#include <pthread.h>
#include <string.h>

#define LIMIT 1000000000
#define NUM_THREADS 8


void format_with_commas(char* out, int n) {
    char buffer[50];
    sprintf(buffer, "%d", n);

    int buffer_len = strlen(buffer);
    int comma_count = (buffer_len - 1) / 3;

    char* p = out;
    for (int i = 0; i < buffer_len; ++i) {
        *p++ = buffer[i];
        if (i < buffer_len - 1 && (buffer_len - 1 - i) % 3 == 0) {
            *p++ = ',';
        }
    }
    *p = '\0';  // Null-terminate the string
}

int digits(int n) {
    if (n < 10) return 1;
    else if (n < 100) return 2;
    else if (n < 1000) return 3;
    else if (n < 10000) return 4;
    else if (n < 100000) return 5;
    else if (n < 1000000) return 6;
    else if (n < 10000000) return 7;
    else if (n < 100000000) return 8;
    else if (n < 1000000000) return 9;
    else return 10;
}

int is_funky_fraction(int denominator) {
    int denominator_digits = digits(denominator);
    int powers_of_10[11] = {1, 10, 100, 1000, 10000, 100000, 1000000, 10000000, 100000000, 1000000000};

    for (int cut_index = 1; cut_index < denominator_digits; ++cut_index) {
        int divisor = powers_of_10[cut_index];
        int reduced_numerator_multiplier = powers_of_10[denominator_digits - cut_index];

        int overlap = denominator / divisor;
        int reduced_denominator = denominator % divisor;

        if (reduced_denominator == 0 || digits(reduced_denominator) + 1 < cut_index) {
            continue;
        }

        int overlap_times_reduced_denominator = overlap * reduced_denominator;
        int denominator_minus_reduced_denominator_times_multiplier = denominator - reduced_denominator * reduced_numerator_multiplier;

        if (overlap_times_reduced_denominator % denominator_minus_reduced_denominator_times_multiplier == 0) {
            int reduced_numerator = overlap_times_reduced_denominator / denominator_minus_reduced_denominator_times_multiplier;
            if (0 < reduced_numerator && reduced_numerator < reduced_denominator && reduced_denominator % reduced_numerator == 0) {
                return 1; // True
            }
        }
    }
    return 0; // False
}

typedef struct {
    int start;
    int end;
    int count;
} ThreadArg;

void* thread_func(void* arg) {
    ThreadArg* thread_arg = (ThreadArg*) arg;
    thread_arg->count = 0;
    for (int denominator = thread_arg->start; denominator < thread_arg->end; ++denominator) {
        thread_arg->count += is_funky_fraction(denominator);
    }
    return NULL;
}

int count_funky_fractions() {
    pthread_t threads[NUM_THREADS];
    ThreadArg thread_args[NUM_THREADS];
    int chunk_size = LIMIT / NUM_THREADS;

    // Create threads
    for (int i = 0; i < NUM_THREADS; ++i) {
        thread_args[i].start = i * chunk_size + 1;
        thread_args[i].end = (i == NUM_THREADS - 1) ? LIMIT : thread_args[i].start + chunk_size;
        pthread_create(&threads[i], NULL, thread_func, &thread_args[i]);
    }

    // Join threads and sum up results
    int total_count = 0;
    for (int i = 0; i < NUM_THREADS; ++i) {
        pthread_join(threads[i], NULL);
        total_count += thread_args[i].count;
    }

    return total_count;
}


int main() {
    struct timespec start, end;
    clock_gettime(CLOCK_MONOTONIC, &start);

    int funky_fraction_count = count_funky_fractions();

    clock_gettime(CLOCK_MONOTONIC, &end);
    double total_time = end.tv_sec - start.tv_sec;
    total_time += (end.tv_nsec - start.tv_nsec) / 1000000000.0;

    char limit_str[50];
    char funky_str[50];

    format_with_commas(funky_str, funky_fraction_count);
    format_with_commas(limit_str, LIMIT);
    printf("%s Funky Fractions up to %s\n", funky_str, limit_str);
    printf("Total time taken: %.2f seconds\n", total_time);

    return 0;
}

File: funky_fractions_multi.py
import time
import multiprocessing


LIMIT = 10_000_000


def digits(n):
    if n < 10:
        return 1
    elif n < 100:
        return 2
    elif n < 1_000:
        return 3
    elif n < 10_000:
        return 4
    elif n < 100_000:
        return 5
    elif n < 1_000_000:
        return 6
    elif n < 10_000_000:
        return 7
    elif n < 100_000_000:
        return 8
    elif n < 1_000_000_000:
        return 9
    else:
        return 10


# Precompute powers of 10
powers_of_10 = [10 ** i for i in range(digits(LIMIT) + 1)]


def is_funky_fraction(denominator):
    denominator_digits = digits(denominator)

    # iterate every possible cut location
    for cut_index in range(1, denominator_digits):
        divisor = powers_of_10[cut_index]
        reduced_numerator_multiplier = powers_of_10[denominator_digits - cut_index]

        # beginning of the denominator that overlaps with the end of the numerator
        overlap = denominator // divisor

        reduced_denominator = denominator % divisor

        # if reduced denominator starts with a zero, skip it
        # if not using strings, we could make sure that the number of digits is expected
        if reduced_denominator == 0 or digits(reduced_denominator) < cut_index:
            continue

        # see if there's an integer solution
        overlap_times_reduced_denominator = overlap * reduced_denominator
        denominator_minus_reduced_denominator_times_multipler = (denominator - (reduced_denominator * reduced_numerator_multiplier))
        is_integer_solution = overlap_times_reduced_denominator % denominator_minus_reduced_denominator_times_multipler == 0

        if is_integer_solution:
            reduced_numerator = overlap_times_reduced_denominator // denominator_minus_reduced_denominator_times_multipler
            # Ensure that the reduced fraction is a positive unit fraction less than 1
            if 0 < reduced_numerator < reduced_denominator and reduced_denominator % reduced_numerator == 0:
                # `numerator` is only used for the print statement
                # numerator = int((reduced_numerator * 10 ** cut_index) + overlap)
                # print(f"{numerator}/{denominator}\t{reduced_numerator_int}/{reduced_denominator}")
                # print(f"{denominator}")
                return True

    return False


def count_funky_fractions():
    with multiprocessing.Pool() as pool:
        results = pool.map(is_funky_fraction, range(11, LIMIT))
    return sum(results)


# Example usage
if __name__ == "__main__":
    start_time = time.time()
    funky_fraction_count = count_funky_fractions()
    end_time = time.time()
    total_time = end_time - start_time
    print(f"{funky_fraction_count} Funky Fractions")
    print(f"Total time taken: {total_time:.2f} seconds")

# Tests
# assert is_funky_fraction(98) == True
# assert is_funky_fraction(64) == True
# assert is_funky_fraction(95) == True
# assert is_funky_fraction(664) == True
# assert is_funky_fraction(110) == True
# assert is_funky_fraction(727040) == True
# assert is_funky_fraction(49450545) == True
# assert is_funky_fraction(22) == False
# assert is_funky_fraction(1001091) == False
